### ISSUE TITLE
Multi cluster support?

### ISSUE THREAD

**Issue by omerlh** [grafana/loki#1866]
Is there an official recommendation regarding running Loki on multiple clusters? Currently, each time we’re deploying a cluster, we need to set up a new data source in Grafana manually (we have single Grafana showing data from all clusters). It’s a bit hard to scale, maintain and use (cannot loon for data from all clusters). 
Are there any existing solutions? It’s not the same issue as tenancy, more like the issue that Thanos/federation solved for Prometheus - one instance that can be used for querying all clusters.

**Comment by cyriltovena** [comment]
Right now the only scale path we offer is tanka in this repository `production/ksonnet/`, it's a distributed cluster where you can send all your logs from any other kubernetes cluster. (That version scales easily to 150MB/sec)

In the future we're looking at improving helm for that same use case and we're also thinking about loki to loki remote writes.

Regarding the datasource you can provision it via the Grafana API, but that doesn't solve looking at data from all cluster.

Federation is a good idea, but I'm not sure the current helm chart can handle multiple cluster load.

**Comment by omerlh** [comment]
We’re using jsonnet to install Loki today, so if federation is possible I’ll be happy to test it.
Where I can read more about the distributed cluster architecture? Does it mean that one cluster is the master, and the rest just write to it?

**Comment by cyriltovena** [comment]
No you just run a single cluster and it can eat all you're sending. It's not federation.

I assume you were using helm, if you already use jsonnet, then as a workaround for now you could just have a single big cluster and ingest all your k8s cluster with it.

**Comment by omerlh** [comment]
So promtail will send all the logs to a centralized loki instance?

**Comment by cyriltovena** [comment]
Yes that’s how we run it.

**Comment by slim-bean** [comment]
Closing this issue as I _think_ this is resolved, please comment if that's not the case. Thanks!

**Comment by tabishis** [comment]
In my case, I've multiple Kubernetes clusters which host apps which specific teams support. In each cluster, I've Loki and Promtail deployed. I've a centralized Grafana from which I create a data source to each cluster's Loki endpoint. It's too many Lokis to manage. What I'd like to do is only have promtail deployed to each Kubernetes cluster which will push data to a centralized Loki instance. Problem with this approach is the centralized Loki instance will contain data for all Kubernetes clusters, which in turn makes all the data available from all K8S clusters to all teams. I'd rather not filter data in Grafana dashboards by namespace for Loki logs to display to each team. Is there a way to control this at Loki level?

**Comment by migueleliasweb** [comment]
I have a very similar case as @omerlh . 

For me, the more important feature would be a single data source to Loki API from `Grafana`, including the ability to aggregate data scattered between different Loki instances.

I don't think there's any documentation regard running Liki in multiple cluster setups =/. Not sure if multi tenancy is quite suitable in this scenario.

**Comment by onhachoe** [comment]
I'm looking into this as well, where we would like to have multiple loki clusters separated geographically, but have a single Grafana for querying aggregate logs.
If it would take long time before any kind of multi-cluster federation support is there, we could use some sort of fanout layer to query multiple clusters and aggregate results in single contained response.

**Comment by slim-bean** [comment]
There are a couple different concerns here and use cases here I can provide some feedback and my opinions on so I'll try to break them out.

Currently the recommended and really only way to use Loki over many clusters is to run a single Loki instance and send logs from all your clusters all over the world to that centralized Loki, this is how we run Loki.  We attach a cluster label at each promtail and are able to query all our logs using Grafana and a single datasource.

> I was looking into the single stretched loki cluster, but wasn't fully sure how perf. would be impacted, so holding off this route.

Turnaround from promtail to logs in Loki is on average less than 2s for all our clusters, Promtail sends batches either when they are full or after a 1s timeout which likely adds to a fair amount of this 2s delay.  I wouldn't expect performance of a centralized Loki cluster to be an issue.

> Cons: All logs are centralized. If something happens to the cluster where the logs are persisted, it's dooms day.

Yes but this is a bit nuanced, as if you run Loki separated in all your clusters, if anything happens to any of your clusters the logs in those clusters are gone, so while you are minimizing the scope of the impact it's not really addressing the problem.  To get multiple region or availability zone durability with Loki you would need to run separate clusters in separate regions and configure promtail to send to both of them (there are some caveats with this approach in how promtail is coded today), or run separate promtail instances for each cluster (probably the best approach).  Or you could look at replicating the chunk and index data between regions by a different mechanism and setup a standby Loki cluster and use some kind of failover DNS, etc.

Mostly you would need to use any of the methods you would use today to make a system available/durable against region or AZ wide issues and I don't think this is really an argument against running a centralized Loki.  That being said, and I'm not even sure what form this would take yet, I'm sure there is more we should do and will hopefully do to make it easier to have better Loki availability over different cloud provider regions.

> Run Loki in it's distributed form (separate components deployed) across multiple clusters. Each cluster would have it's down instance of distributor, ingester and querier. This would require some level of orchestration to ensure all components are properly configured at all times - maybe a small operator would be necessary.

This could be made possible, the distributed hash ring used by Loki could probably be made location aware to make sure replicated data is sent to different locations but this kind of setup would work very poorly for querying and I don't think would be a good fit for Loki without some big changes into how the querying works.  I think solving the HA over multiple regions can probably be solved in other ways a bit easier in other ways.

> In my case, I've multiple Kubernetes clusters which host apps which specific teams support. In each cluster, I've Loki and Promtail deployed. I've a centralized Grafana from which I create a data source to each cluster's Loki endpoint. It's too many Lokis to manage. What I'd like to do is only have promtail deployed to each Kubernetes cluster which will push data to a centralized Loki instance. Problem with this approach is the centralized Loki instance will contain data for all Kubernetes clusters, which in turn makes all the data available from all K8S clusters to all teams. I'd rather not filter data in Grafana dashboards by namespace for Loki logs to display to each team. Is there a way to control this at Loki level?

I think this is an interesting question and there is probably a good feature request for Grafana in here which is provisioning a Loki datasource and specifying certain labels that are require for each query.  This is already possible in a different way today just using dashboards, you can make a dashboard with a log panel and template variables for changing labels, but in the query editor you can hard-code certain labels.  Take a look at [this blog](https://grafana.com/blog/2020/04/08/loki-quick-tip-how-to-create-a-grafana-dashboard-for-searching-logs-using-loki-and-prometheus/) for some more details.

I'm not sure if there are other advantages or reasons to do it at a datasource level or not but I hadn't though of that before.

_some additional comments:_

First thanks for the discussion and feedback everyone! This is an area that clearly needs some better documentation, guidance and work.

TL;DR Multi cluster Loki support is best obtained via a single Loki instance and remote agents sending with a cluster specific label.  HA for Loki over regions is currently going to be best accomplished by running separate Loki instances and sending data to both or replicating the Loki storage data (chunks and index) via another means to a different region to a standby Loki cluster.

In related features though I do think we want to support a federation model in the future of some kind with Loki where you could run a Loki instance per cluster and also have that instance send data to a centralized Loki instance (either all the data or a subset).

**Comment by migueleliasweb** [comment]
Thank you so much, @slim-bean ! That was some very useful ideas! 

I will be spiking one of theses setups in following week. Cheers!

**Comment by lipmanp** [comment]
The actual query will be distributed to nodes, and results will be aggregated back in query client. it would be similar to how Elasticsearch query works.

**Comment by yanirj** [comment]
Regarding @lipmanp's comment - is it possible to run a single hub Loki query frontend (https://grafana.com/docs/loki/latest/configuration/query-frontend/) and have all other clusters' Loki instances point to the hub frontend? (let's assume multi cluster service discovery works or a LoadBalancer is in place)

I'm asking because I'm not sure if having answers back from multiple Loki instances would be really aggregated or the original query will simply be split to N parts of different time frames will be sent to each Loki instance. I'm not interested in parallelization here to accelerate queries (query frontend's original goal), just aggregation. Basically similar to a Thanos setup.

**Comment by tivvit** [comment]
> Regarding @lipmanp's comment - is it possible to run a single hub Loki query frontend (https://grafana.com/docs/loki/latest/configuration/query-frontend/) and have all other clusters' Loki instances point to the hub frontend? (let's assume multi cluster service discovery works or a LoadBalancer is in place)

I don't think that it is possible now but I am thinking about this solution for a while. I will try to implement some POC over the weekend.

**Comment by lipmanp** [comment]
regarding to @slim-bean comments on how to achieve HA, it seems only 2 ways: 
1) to run separate Loki instances in parallel or 
2) set up a mechanism to replicate index/chunks to another location.
but in practice, I still need some guidance:
for 1), you could switch data source to the healthy Loki instance, but the question is how to rebuild the failure one and having data sync with the working instance
for 2), I'm not sure if index/chunks  always contain latest data, e.g. data in memory within working instance is in sync with physical index/chunks files?

without recovering quickly, only one Loki instance running, no another instance to be switched to, if it is failed luckily.

**Comment by testfile2** [comment]
Has this gone anywhere? As @lipmanp pointed out there are some questions with the point 2. 

If I can throw another question in the loop here: What if you share the store? As in have 1 S3 bucket that multiple clusters/loki instances point to? Then there is no need for replicating the chunks over? Ofc this still falls apart from @lipmanp point 2.2 but I guess thats the only way to do this for now? Until we have native support for this kind of thing.

**Comment by DevSusu** [comment]
Curious about @testfile2 's question!

Wonder if there's any method of combining index & data storage

BTW sending logs to local Loki + central Loki also sounds promising! Thanks for participating in this thread with real-world benchmarks @slim-bean :+1:

**Comment by pourquoi42** [comment]
@testfile2 @cyriltovena 
I am also really curious about the approach @testfile2 proposed:
> What if you share the store? As in have 1 S3 bucket that multiple clusters/loki instances point to? Then there is no need for replicating the chunks over?
In our scenario, one k8s cluster is not enough to host all workloads. It reaches limits on resources/namespaces and load on api-server. Therefore we have multiple clusters in the same region for horizontal scale-out of workloads. Having a central loki and all promtails sending to it is not desirable for several reasons for us. The "shared storage idea" would be ideal.

I've read up the documentation and it seems to me, that this would work out of the box in the write path. For the read path, there is a problem because the indices are flushed to S3 only every 15 min. Local loki therefore also queries the ingestors. However, the problem that queries run over 15 min old logs is not a big deal for us in most cases. For the few cases where it matters, there are ways to find the right cluster to ask . 

What am I missing? Can anyone who knows the implementation in detail see a problem why this shouldnt work?

**Comment by hauks96** [comment]
What about the following approach:

- Use one S3 bucket for storage
- Use one `tenant_id` per cluster, f.x `dev-cluster` & `staging-cluster`
- Loki instance on each cluster only querying data for it's given tenant_id with `multi_tenant_queries_enabled: false`
- Observer cluster running only loki (not promtail) with all tenant_id's piped in the `X-Scope-OrgID` header such as  
  `dev-cluster | staging-cluster`

Is this something that could work?

**Comment by AlissonRS** [comment]
Hello @slim-bean ! Would you have anything to add to @testfile2 suggestion? Also about @pourquoi42's comment.

I really don't want a centralized loki cluster, that's one extra cluster to manage just for the sake of Loki, seems overkill to me, not to mention it's a single point of failure, if the Loki cluster goes down I lose my ability to inspect logs while it.

Ideally I'd have the whole Loki stack running on two clusters, sharing same S3 buckets.

I also have Grafana running on the two clusters, and my goal would be that no matter if I open the Grafana UI from cluster A or B, it'd show the Loki logs from both clusters.

This way, if one cluster goes down, I can still access logs from Grafana in the other cluster, the only issue would be losing potential logs that weren't flushed yet from the cluster that went down, but it's not a big deal to me.

**Comment by sourcehawk** [comment]
@AlissonRS 

- Deploy promtail on each cluster. For the sake of this example I'll name the clusters `tenant_1` and `tenant_2`. In the promtail helm chart there is a variable called `tenant_id` which you can use to identify logs from the clusters when promtail pushes them to loki.
- Configure the loki instance on each cluster to write to the same S3 bucket and enable the flag `multi_tenant_queries_enabled` on the loki querier
- On both your clusters, create the loki grafana datasource with an `X-Scope-OrgID` header set to `tenant_1 | tenant_2` 

Now both of your clusters will query data from both tenants. Note that if you for instance set `X-Scope-OrgID` to `tenant_1` on cluster 1, you would only query data from the cluster that has promtail configured with `tenant_1`

**Comment by outofrange** [comment]
@sourcehawk wouldn't this have the drawback that logs are only shipped to S3 every `max_chunk_age` seconds/minutes/hours?

Looking at [Stackoverflow](https://stackoverflow.com/a/78685891), people configure this as low as `30s` - which doesn't sound like the best idea if you don't want to produce tons of blocks.
But with the default of `2h`, you would only see recent logs from the Loki you're actually querying, not from other clusters.

But to solve querying from multiple Loki instances, you'd need something like _Thanos Query_ (as mentioned earlier), distributing queries to all instances / stores.

**Comment by sourcehawk** [comment]
@outofrange Right, completely forgot about that part. Seems like the only viable option here would be to expose the loki api on each cluster and make promtail on each cluster push the logs to all of your clusters. 

I.e you would need to expose the loki api on each clusters and the configure promtail like this:
```yaml
promtail:
    clients:
      # local cluster
      - url: http://loki-gateway/loki/api/v1/push
        tenant_id: "cluster-1"
     # external clusters
      - url: https://loki-gateway.cluster-2.com/loki/api/v1/push
        tenant_id: "cluster-1"
      - url: https://loki-gateway.cluster-3.com/loki/api/v1/push
        tenant_id: "cluster-1"
```

On cluster 2 you would have 
```yaml
promtail:
    clients:
      # local cluster
      - url: http://loki-gateway/loki/api/v1/push
        tenant_id: "cluster-2"
     # external clusters
      - url: https://loki-gateway.cluster-1.com/loki/api/v1/push
        tenant_id: "cluster-2"
      - url: https://loki-gateway.cluster-3.com/loki/api/v1/push
        tenant_id: "cluster-2"
```
and so forth.

However you might end up with `n` duplicates of the same log in your s3 bucket unless loki has a deduplication mechanism that removes exact duplicates of logs (identical log and timestamp). `n` would be the number of clusters you are managing
